### ISSUE TITLE
Library/Dictionary as Token?

### ISSUE THREAD

**Issue by JAlflen** [gurnec/btcrecover#97]
Is it possible to include a library/dictionary in place of a token to eliminate having to try every combination of letters?

For example, say I know my password begins with "Cold" and ends with "45st" and I know between them exists some actual word in an english dictionary but I have no clue as to what it even begins with.

**Comment by gurnec** [comment]
Which wallet type is this for (what software created your wallet)? Which OS are you running on?

[John the Ripper](https://github.com/magnumripper/JohnTheRipper) is probably a better tool for this task, assuming it has support for your wallet type. I could give you some pointers on how to use it.

Otherwise, writing a small script to create a big tokens or a dictionary file would be possible. Is the tokens file your thinking of exactly as simple as you described, or were you just giving an example / is it actually more complicated than what you described?

**Comment by JAlflen** [comment]
Hi,

This is for a blockchain.info wallet. I am using Ubuntu as my OS. 

Yes my password really is as simple as I described. Knowing this, do you still recommend John the Ripper over btcrecover?

Thanks,
Joe

**Comment by gurnec** [comment]
Yes I'd recommend JtR over btcrecover in your case. It has good blockchain.info support, and it's not difficult to compile on Ubuntu.

 * Download/build instructions: <https://github.com/magnumripper/JohnTheRipper/blob/bleeding-jumbo/doc/INSTALL-UBUNTU>
 * Use `blockchain2john.py` in the run directory to extract a "hash" from your wallet and store it in a hashes.txt file.
 * Find a dictionary file to use, Top95Thousand-probable.txt [from here](https://github.com/berzerk0/Probable-Wordlists/tree/master/Real-Passwords) might be a good start, and bigger ones are available from there too.
 * Run john: `./john --wordlist=Top95Thousand-probable.txt --mask='Cold?w45st' hashes.txt`

**Comment by JAlflen** [comment]
Thank you so much for your help! I will definitely provide a donation upon my success.

**Comment by gurnec** [comment]
Thanks, and good luck!

**Comment by JAlflen** [comment]
Question, in this command:
./john --wordlist=Top95Thousand-probable.txt --mask='Cold?w45st' hashes.txt

Is the mask essentially what I know of my password and the '?' is replaced
with a word from the file?
If so, is it possible to capitalize the first letter in the wordlist?

Thanks

**Comment by JAlflen** [comment]
Nevermind, figured it out. Sorry.

**Comment by JAlflen** [comment]
Okay I thought I had it but I think there is an issue with step 2. How
exactly do I
-Use blockchain2john.py in the run directory to extract a "hash" from your
wallet and store it in a hashes.txt file

I was just doing blockchain2john.py wallet.aes.json --json > hashes.txt
but am recieving 0 password hashes cracked, 1 left with an empty john.pot
file.

**Comment by gurnec** [comment]
You said you already figured this out, but I'm just double-checking...

> Is the mask essentially what I know of my password and the '?' is replaced
> with a word from the file?

The `?w` is replaced with a word from the file. See the `doc/MASK` file for details.

> is it possible to capitalize the first letter in the wordlist?

Yes, but you have to use the more complicated "rules" mode instead of the "mask" mode.

Create a file named `john-local.conf` in the same directory as the john binary, and add this:
```
[List.Rules:MyRules]
A0"Cold" Az"45st"
(?x T0 A0"Cold" Az"45st"
```
The first line prefixes and appends the two strings to each word from the file (the `A0` and `Az` rules), and the second line only runs if the first character is a letter (the `(?x` rule) which it then swaps the case of (the `T0` rule) and then it does the prefixing and appending again. See the `doc/RULES` file for details.

They you run john like: `./john --wordlist=Top95Thousand-probable.txt --rules=MyRules hashes.txt`

> I was just doing blockchain2john.py wallet.aes.json --json > hashes.txt
> but am recieving 0 password hashes cracked, 1 left with an empty john.pot
> file.

That looks correct. The "1 left" message means that it understands the hash that's in the hashes.txt file, but your combination of wordlists and/or masks and/or rules didn't find the correct password. You may need a bigger wordlist file, or you may be wrong about the prefix/postfix you're using.

[This file](https://github.com/berzerk0/Probable-Wordlists/raw/master/Dictionary-Style/MainEnglishDictionary.txt) (right-click, Save as...) is the next bigger wordlist from [that repo](https://github.com/berzerk0/Probable-Wordlists) I posted above, and could work well with the MyRules from above. The other even bigger wordlists are available as torrents in that repo.

**Comment by JAlflen** [comment]
I got it all working and cracked a test wallet. I thought it was a little easier to just use sed to create a whole new word list with the first letter capitalized.

Thanks again.

**Comment by gurnec** [comment]
That works too 😃

**Comment by nori19870522** [comment]
I am Japanese and do password analysis.
https://keychainx.io/ 

This is Robert's article.
https://news.bitstarz.com/bitstarz-news-interview-robert-rhodin-keychainx-ceo  

I may be able to find the password you lost. Do you need help with your login?

**Comment by noobcodeer90** [comment]
Hey gurnec how can I crack my second wallet password assuming that knowing only first few words of my password, also I used some special characters in between. 
Is that even possible to crack with johntheripper in windows?
